File: LeetCode/150.evaluate-reverse-polish-notation.cpp
/*
 * @lc app=leetcode id=150 lang=cpp
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        stack<int> stack;
        for (string &s : tokens)
            if (s == "+" || s == "-" || s == "*" || s == "/")
            {
                int op1, op2, res;
                op1 = stack.top();
                stack.pop();
                op2 = stack.top();
                stack.pop();
                if (s == "+")
                {
                    res = op2 + op1;
                }
                if (s == "-")
                {
                    res = op2 - op1;
                }
                if (s == "/")
                {
                    res = op2 / op1;
                }
                if (s == "*")
                {
                    res = op2 * op1;
                }
                stack.push(res);
            }
            else
            {
                stack.push(stoi(s));
            }
        return stack.top();
    }
};
// @lc code=end

File: LeetCode/150.evaluate-reverse-polish-notation.py
#
# @lc app=leetcode id=150 lang=python3
#
# [150] Evaluate Reverse Polish Notation
#


# @lc code=start
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        def evaluate(op1, op2, operator):
            match operator:
                case "+":
                    return op1 + op2

                case "-":
                    return op1 - op2

                case "*":
                    return op1 * op2

                case "/":
                    return int(op1 / op2)

        stack = []

        for token in tokens:
            if token in "+-*/":
                op2 = stack.pop()
                op1 = stack.pop()
                stack.append(evaluate(op1, op2, token))

            else:
                stack.append(int(token))

        return stack.pop()


# @lc code=end
